awscli/shorthand.py

```diff
@@ -282,3 +332,71 @@ def _backtrack_to(self, char):
     def _consume_whitespace(self):
         while self._current() != _EOF and self._current() in string.whitespace:
             self._index += 1
+
+
+class ModelVisitor(object):
+    def visit(self, params, model):
+        self._visit({}, model, '', params)
+
+    def _visit(self, parent, shape, name, value):
+        method = getattr(self, '_visit_%s' % shape.type_name,
+                         self._visit_scalar)
+        method(parent, shape, name, value)
+
+    def _visit_structure(self, parent, shape, name, value):
+        if not isinstance(value, dict):
+            return
+        for member_name, member_shape in shape.members.items():
+            self._visit(value, member_shape, member_name,
+                        value.get(member_name))
+
+    def _visit_list(self, parent, shape, name, value):
+        if not isinstance(value, list):
+            return
+        for element in value:
+            self._visit(value, shape.member, '', element)
+
+    def _visit_map(self, parent, shape, name, value):
+        if not isinstance(value, dict):
+            return
+        value_shape = shape.value
+        for k, v in value.items():
+            self._visit(value, value_shape, k, v)
+
+    def _visit_scalar(self, parent, shape, name, value):
+        pass
+
+
+class BackCompatVisitor(ModelVisitor):
+    def _visit_list(self, parent, shape, name, value):
+        if not isinstance(value, list):
+            # Convert a -> [a] because they specified
+            # "foo=bar", but "bar" should really be ["bar"].
+            if value is not None:
+                parent[name] = [value]
+        elif shape.member.type_name == 'structure' and \
```

Comment:
I think this can actually be removed.  The visitor was written before plumbing it into the argprocess module.  Unfortunately the visitors are applied *after* parsing, so there's a requirement that the shorthand parser can interpret `i-123 i-124 i-125` and parse it into a list of strings: `['i-123', 'i-124', 'i-125']`.  This would complicate the parser to add additional special cased grammar rules.

So I opted to instead handle this special case in the argprocess module (via `_handle_special_cases`).  This means this processing can be removed from the visitor class because it's relying on an input format that the parser will never produce.